src/ddmd/link.d

```diff
@@ -1039,4 +983,384 @@ version (Windows)
+                    VCInstallDir = keyVC.GetString("ProductDir");
+                    if (VCInstallDir)
+                        break;
+                }
+        }
+
+        void detectVCToolsInstallDir()
+        {
+            if (VCToolsInstallDir is null)
+                if(char* pe = getenv("VCTOOLSINSTALLDIR"))
+                    VCToolsInstallDir = pe;
+
+            if (VCToolsInstallDir is null && VCInstallDir)
+            {
+                const(char)* defverFile = FileName.combine(VCInstallDir, r"Auxiliary\Build\Microsoft.VCToolsVersion.default.txt");
+                if (FileName.exists(defverFile))
+                {
+                    // VS 2017
+                    File f = File(defverFile);
+                    if (!f.read()) // returns true on error (!), adds sentinel 0 at end of file
+                    {
+                        auto ver = cast(char*)f.buffer;
+                        // trim version number
+                        while (*ver && isspace(*ver))
+                            ver++;
+                        auto p = ver;
+                        while (*p == '.' || (*p >= '0' && *p <= '9'))
+                            p++;
+                        *p = 0;
+
+                        if (ver && *ver)
+                            VCToolsInstallDir = FileName.combine(FileName.combine(VCInstallDir, r"Tools\MSVC"), ver);
+                    }
+                }
+            }
+        }
+
+        enum VCDir { Base, Bin, Lib }
+
+        const(char)* getVCDir(VCDir dir, bool x64)
+        {
+            if (VCToolsInstallDir !is null)
+            {
+                if (dir == VCDir.Bin)
+                    return FileName.combine(VCToolsInstallDir, x64 ? r"bin\HostX86\x64" : r"bin\HostX86\x86");
+                if (dir == VCDir.Lib)
+                    return FileName.combine(VCToolsInstallDir, x64 ? r"lib\x64" : r"lib\x86");
+                return VCToolsInstallDir;
+            }
+            if (dir == VCDir.Bin)
+                return FileName.combine(VCInstallDir, x64 ? r"bin\amd64" : "bin");
+            if (dir == VCDir.Lib)
+                return FileName.combine(VCInstallDir, x64 ? r"lib\amd64" : "lib");
+            return VCInstallDir;
+        }
+
+        const(char)* getUCRTLibPath(bool x64)
+        {
+            if (UCRTSdkDir && UCRTVersion)
+               return PathBuilder(UCRTSdkDir) / "Lib" / UCRTVersion / (x64 ? r"ucrt\x64" : r"ucrt\x86");
+            return null;
+        }
+
+        const(char)* getSDKLibPath(bool x64)
+        {
+            if(WindowsSdkDir)
+            {
+                const(char)* arch = x64 ? "x64" : "x86";
+                auto sdk = PathBuilder(WindowsSdkDir) / "lib";
+                if(FileName.exists(sdk / WindowsSdkVersion / "um" / arch / "kernel32.lib")) // SDK 10.0
+                    return sdk / WindowsSdkVersion / "um" / arch;
+                else if(FileName.exists(sdk / r"win8\um" / arch / "kernel32.lib")) // SDK 8.0
```

Comment:
I'd rather see `a / b / c` added as a function to `root.filename`, so it can be used elsewhere. It should use variadic parameters to list them, or an array literal, or even 2,3,4 parameter overloads.

Reply:
I introduced the PathBuilder struct because FileName.combine cannot be used with UFCS, but creates unreadable code if nesting multiple calls.
Using variadic parameters instead is almost as good, though, so I replaced PathBuilder with FileName.buildPath now.